### ISSUE TITLE
Suggestion/Feature request: Regex (and shortcuts) for getWindowsWithTitle's title

### ISSUE THREAD

**Issue by macdeport** [Kalmat/PyWinCtl#8]
Regex for generality and versatility but shortcuts `contains`, `endswith`, `is`, `matches` (regex), `startswith` and negative counter part for comfort in common situations.

**Comment by Kalmat** [comment]
Hi! 

Do you mean adding these options when calling getWindowsWithTitle()? In such a case, what do you have in mind or you think it would be more convenient? Maybe using notations, something like this:

```
'Chrome' - matches "Chrome"
'Chrome\*' - startswith "Chrome"
'\*Chrome' - endswith "Chrome"
'\*Chrome\*' - contains "Chrome"
```

Do you miss any additional option?

Apart from this, I am not sure to understand the "negative counter part" concept, sorry. Could you please explain?

**Comment by macdeport** [comment]
`'Chrome' - matches "Chrome" => 'Chrome' - is "Chrome"`

The "negation" counter part:
`'Chrome' - is not "Google"`

`'Chrome' - does not endswith "Ome"`

etc

**Comment by Kalmat** [comment]
Understood. Thanks for clarifying. Let me think how this can be done without getting too complex or... are you thinking to use all options and notations included in [re module](https://docs.python.org/3/library/re.html)? This last option is powerful in deed, but extremely complex and not sure if everyone is able to use it with ease (I'm not!!! HAHAHAHAHA!)

I would suggest to add the "counter part" options, and a case-sensitive (True / False) argument... what do you think?

**Comment by macdeport** [comment]
I suggest something like this (without test from side):

```
getWindowsWithTitle(title,app='all',condition=['contains'|'endswith'|'is'|'matches'|'startswith'|
                                                  'does not contain'|'does not endswith'|'is not'|
                                                  'does not match'|'does not startswith'])

...
condition=condition.strip().lower()
if condition=='contains':
  cond_func=lambda s1,s2: s1 in s2
elif condition=='endswith':
  cond_func=lambda s1,s2: s2.endswith(s1)
...
else:
   Error!

```
:-)

**Comment by macdeport** [comment]
> DId you see the last addition: WinWatchDog? I think it can be very useful. Let me know what you think whenever you have the chance to check it.

Sure WinWatchDog can be useful... But where is it hiding? Do you have a test program?

I like your PS code but dreams needs to be living.

**Comment by Kalmat** [comment]
HAHAHA! It's not hiding! It's... forgotten. I mean, I forgot to push it!!! I will upload it together with the new getWindowsWithTitle() function. Sorry!

PS: `if "dream" in List(things_I_want): subprocess.Popen("make it_happen -B -i -k")`

**Comment by Kalmat** [comment]
Hi again! I uploaded a preliminary version including WatchDog, but I decided it was better to encapsulate it. Don't test it by now if you had the intention. Sorry for the inconveniences. The getWindowsWithTitle() is ready to check.

Thanks!

**Comment by Kalmat** [comment]
I think it's all ready now. Just to let you know. Of course, no hurry nor any obligation to do anything with it. Anyway, bear in mind that Apple Script version of watchdog (as in general) might be very slow!

**Comment by macdeport** [comment]
https://github.com/Kalmat/PyWinCtl/issues/8#issuecomment-1083532779
I took a look at the code and there must be 2 bugs:
1. You don't have to change the "title" case if it's a regex (in case of regex the case management depends on `re.IGNORECASE` & `re.LOCALE` i.e. regex_flags value)
2. confusion about how `re.match` and `re.search` works

**Comment by Kalmat** [comment]
Hi!

1. You're right!
2. The functionality, syntax and differences are explained in the links provided in the function docstring (https://docs.python.org/3/library/re.html). Furthermore, I assumed that people using RegEx would have knowledge of both. I could've chosen one or the other (match vs search), but I didn't want to prevent devs from having that option since both are useful depending on the needs. If I should choose just one, that would be .search(), which is more general.

Let me please know your thoughts

**Comment by macdeport** [comment]
Suggestion
As you have fixed conditions the coding can be more pythonic using direct addressing e.g. key => value in place of if / elif

**Comment by Kalmat** [comment]
Hi again!

1. Sorry, in which sense is not OK?
2. Agree. To be honest, I included match because you mentioned. I also think search is far better.
3. "key =>value"... I'm affraid I don't understand. There is no dictionary in this function. Can you please explain?

Thanks!

**Comment by macdeport** [comment]
1.
 ```Python

def _addWin(winTitle, title, condition, regex_flags):
    addWin = False
    if condition == Re.RegExMatch:
        addWin = re.**search**(title, winTitle, regex_flags) is not None
    elif condition == Re.RegExSearch:
        addWin = re.**search**(title, winTitle, regex_flags) is not None

```
2. I have used "matches" for string s1 matches s2 , not for `re.match()` ;-)
3. For example:

```
from re import search
cond_dic={'Contains': lambda s1,s2,rgxfl: s1 in s2,
          'EndsWith': lambda s1,s2,rgxfl: s2.endswith(s1),
          'Is': lambda s1,s2,rgxfl: s1==s2,
          'Match': lambda s1,s2,rgxfl: bool(search(s1,s2,rgxfl)),
          'StartsWith': lambda s1,s2,rgxfl: s2.startswith(s1),
          'NotContains': lambda s1,s2,rgxfl: not cond_dic['Contains'](s1,s2,rgxfl),
          'NotEndsWith': lambda s1,s2,rgxfl: not cond_dic['EndsWith'](s1,s2,rgxfl),
          'NotIs': lambda s1,s2,rgxfl: not cond_dic['Is'](s1,s2,rgxfl),
          'NotMatch': lambda s1,s2,rgxfl: not cond_dic['Match'](s1,s2,rgxfl),
          'NotStartsWith': lambda s1,s2,rgxfl: not cond_dic['StartsWith'](s1,s2,rgxfl)
          }

print(cond_dic.keys());regex_flags=0
print(cond_dic['Contains']('56','256',regex_flags))
print(cond_dic['Match']('[652]{3}','256',regex_flags))
print(cond_dic['NotContains']('56','256',regex_flags))

```

**Comment by Kalmat** [comment]
1. Right! Only in macOS part! Thank you for pointing it out!
2. HAHAHAHA! Ok. Using .search() only
3. Nice! Definitely more pythonic and definitely far beyond my python knowledge

Thank you!!!

**Comment by macdeport** [comment]
3. Part of your very near future knowledge :-)

**Comment by Kalmat** [comment]
3. HAHAHAHAHA! Totally right!!!

Well, I think it's working now. I will update the repository. WatchDog encapsulated version is also included. Whenever you try it, bear in mind that Terminal is not suitable to test this, since it changes the name (e.g. 'pywinctl - python - 80x24' vs. 'pywinctl - osascript - 80x24' when an Apple Script is running... so, it does not identify it as the "active" window). Yes, using title is a nightmare, but found no other way.

**Comment by macdeport** [comment]
`WatchDog` is the subject of this issue but I reply ;-) Why not to consider regex difficult watchdogging ?

**Comment by macdeport** [comment]
`getWindowsWithTitle` seems to work like a charm on my side :-)

Nevertheless, the `REGEXSEARCH` condition should be shortened to `MATCH`, in common use in the computer community and even in the Python `re` module documentation... :-) It doesn't matter if it is implemented by `re.search()` internally.

The possible integer `flags` (sum of separate binary flags) companion of `MATCH` seems to me more elegant to provide in the form `condition=MATCH` / `condition=(MATCH, flags)` and makes `:param regex_flags` disappear.

Still about these "comfort" conditions: MATCH generalizes CONTAINS; it would be good for dummies or near-dummy regexes sometimes to add `NOTMATCH` which generalizes `NOTCONTAINS`: not all users will think of using "`(?!...)` Regex zero-width Negative lookahead assertion" to encapsulate their regex if necessary, myself included of course).

**Comment by Kalmat** [comment]
Happy to hear that!

1. Totally fine. I will rename it to `MATCH`
2. I like that suggestion. In fact, I wasn't happy with a param which applies to just one value. Besides, the same thing actually happens to caseSensitive. In this last case, it should be either `condition=(AnyConditionButMATCH, caseSensitive)` or... if caseSensitive is set to False, add `re.IGNORECASE` to regex_flags if it wasn't (`regex_flags = regex_flags | re.IGNORECASE`)?
3. `NOTMATCH` would be just the negation of `MATCH`, right? (`not pattern.search(title)`)

PS: I didn't understand a word of this: "(?!...) Regex zero-width Negative lookahead assertion" HAHAHAHAHA! Sorry for my lack of regex skills....

**Comment by Kalmat** [comment]
Regarding WatchDog, your suggestion sounds interesting. Do you mean using regex to identify the title in case it changes, or using regex to identify the windows you want to monitor?

The first point, may lead to a different window, with a similar title, but not the one we want to monitor. There was a version of `getWindowsWithTitle()` some months ago, which was using levenshtein algorithm (sort of fuzzy comparison which uses a percentage of similarity to compare two strings). I finally discarded it because of that reason, but could be a good option to add now we have other options too... what to do you think? (it's a totally different approach than regex, which can be also useful for some devs, not sure)

The second point could be easily done this way:

```
windows = []
watchdogs = []
titles = getWindowsWithTitle(whatever_condition)
for title in titles:
    wins = getWindowsWithTitle(title)
    for win in wins:
        windows.append[win]
        watchdogs.append[win.watchdog.start()]
```

Do you mean to encapsulate this code within a separate function? Take into account that you are creating window objects and, more "costy", new threads for each watchdog... Maybe the dev should control it by its own. Not sure!

**Comment by macdeport** [comment]
https://github.com/Kalmat/PyWinCtl/issues/8#issuecomment-1086721554
1. 👍
2. Your are right.  For syntax consistency, condition=any except `MATCH`/`NOTMATCH` / `condition=(`any except `MATCH`/`NOTMATCH, caseSensitive)` seems well.
3. You are perfectly right. I have just a thinking about regex negation (please search  "negative lookahead assertion" in `re` documentation" to be in the picture ;-) ).

**Comment by macdeport** [comment]
https://github.com/Kalmat/PyWinCtl/issues/8#issuecomment-1086724636
Regarding `WatchDog`, the dreamer (i have no enough time yet to dig in) just suggest using regex versatility, if needed, to identify the title in case it changes if it change in a quiet predictable way.

**Comment by Kalmat** [comment]
I definitely love this kind of conversation. It helps me a lot!!!

1. Done!
2. I had another idea: using flags as separate param, but with a different meaning according to condition value:
    - MATCH / NOTMATCH: regex flags as per re module
    - LEVENSHTEIN: similarity percentage threshold (0-100)
    - ALL OTHERS: re.IGNORECASE (this is key to elliminate caseSensitive param)
   Thus, the function definition would be: `getWindowsWithTitle(title, app=None, condition=Re.IS, flags=0)`
3. Done! NOTMATCH is the negation of MATCH (both based on regex patterns)

Regarding the watchdog:
 - In AppleScript, not all windows have the attribute "id", which could be ideal for this.
 - Using Quartz, kGCWindowNumber does not match any other value you can get from Apple Script (besides, it changes with some events, like maximizing the window). From Catalina+, Apple removed kGCWindowName. 
 In short, there is no value you can use to be sure you are addressing the right window but the title. 
- `wID` is a mere sequence number I am assigning, no relation to any "real" id on the system.
- `getHandle`, for the AppScript version returns... the title!! (because it's in the end the ID I had to use).
We can use a similarity check for the titles of the open windows within the same running app. This constrains quite much the search, but not enough. E.g. Within "Terminal", if we compare "PyWinCtl - python - 80x24" and "PyWinCtl - osascript - 80x24", they have a similarity of 67%... too low!!!! This is why I discarded the levenshtein method.

I have to admit this is something that is driving me crazy trying to figure out a good solution.

**Comment by Kalmat** [comment]
Sorry, one more thing about this. I was also considering using window bounds (size and position) together with similarity and app id... In other words, if it's the same app AND same size and position AND title is similar (we should define what "similar" means in terms of percentage), we could consider it's the same window... but still too weak.

**Comment by macdeport** [comment]
https://github.com/Kalmat/PyWinCtl/issues/8#issuecomment-1086842563
`getWindowsWithTitle`
I agree globally but don't really understand `re.IGNORECASE` but I understand your try ;-)
Proposal: new param: FIRST window of an app. Perhaps a new wheel to have a synthetic view of .__doc__

**Comment by Kalmat** [comment]
Sorry. Perhaps the function itself with its doc is clearer:

```
Re.IGNORECASE = re.IGNORECASE

def getWindowsWithTitle(title, app=None, condition=Re.IS, flags=0):
    """
    Get the list of Window objects whose title match the given string using the give condition and flags.
    Use ''condition'' to delimit the search. Allowed values are stored in pywinctl.Re sub-class (e.g. pywinctl.Re.Contains)
    Use ''flags'' to define additional values according to each condition type:

        - IS -- window title is equal to given title (allowed flags: Re.IGNORECASE)
        - CONTAINS -- window title contains given string (allowed flags: Re.IGNORECASE)
        - STARTSWITH -- window title starts by given string (allowed flags: Re.IGNORECASE)
        - ENDSWITH -- window title ends by given string (allowed flags: Re.IGNORECASE)
        - NOTIS -- window title is not equal to given title (allowed flags: Re.IGNORECASE)
        - NOTCONTAINS -- window title does NOT contains given string (allowed flags: Re.IGNORECASE)
        - NOTSTARTSWITH -- window title does NOT starts by given string (allowed flags: Re.IGNORECASE)
        - NOTENDSWITH -- window title does NOT ends by given string (allowed flags: Re.IGNORECASE)
        - MATCH -- This means ''title'' param contains a regex pattern (allowed syntax and flags: ''re.search()'' rules at https://docs.python.org/3/library/re.html)
        - NOTMATCH -- ''title'' as a regex pattern too, but returning windows which do NOT fulfill the condition (allowed syntax and flags: ''re.search()'' rules at https://docs.python.org/3/library/re.html)
        - LEVDISTANCE -- Uses levenshtein distance to get a similarity percentage (allowed flags: 0-100 as similarity threshold. Defaults to 90)

    :param title: title or regex pattern to match, as string
    :param app: used ONLY in macOS platform
    :param condition: (optional) condition to apply when searching the window. Defaults to ''Re.IS'' (is equal to)
    :param flags: (optional) flags to apply to condition, according to its value, as described above.
    :return: list of Window objects
    """
    matches = []
    if title and condition in Re._cond_dic.keys():
        lower = False
        if condition in (Re.MATCH, Re.NOTMATCH):
            title = re.compile(title, flags)
        else:
            if condition == Re.LEVDISTANCE and (not isinstance(flags, int) or not (0 < flags <= 100)):
                flags = 90
            if flags & Re.IGNORECASE:
                lower = True
                title = title.lower()
        for win in getAllWindows():
            if win.title and Re._cond_dic[condition](title, win.title.lower() if lower else win.title, flags):
                matches.append(win)
    return matches
```

I will upload a new version fixing the getMenu() method and with this new getWindowsWithTitle() version if you agree.

Thanks!
